### PR TITLE
Refactor datetime-diff implementations to use multimethod instead of case statement

### PR DESCRIPTION
**Pull request by calherries**
This PR refactors a couple of datetimeDiff implementations in master according to Cam's suggestions here https://github.com/metabase/metabase/pull/27064#discussion_r1047758067

Review should start at [src/metabase/driver/sql/query_processor.clj](https://github.com/metabase/metabase/pull/27237/files#diff-de45e1639c2976212706ab34736e8fbca40b497026adca37033da380fca04738), and move to the specific implementations.

### CODE REVIEW

File: src/metabase/util/honeysql_extensions.clj
Comment (calherries) [inline src/metabase/util/honeysql_extensions.clj:221]: Just added a helper function which is called by `datetime-diff-check-args`.
https://github.com/metabase/metabase/pull/27237/files#diff-de45e1639c2976212706ab34736e8fbca40b497026adca37033da380fca04738R684

File: src/metabase/driver/sql/query_processor.clj
Comment (calherries) [inline src/metabase/driver/sql/query_processor.clj:699]: I _think_ this regex can cover the types of all SQL implementations. If not, an individual implementation can modify it. 

Ideally I would have used a function like `hx/is-of-base-type` that can determine whether a field or value is of a certain base type. But sadly that's a little tricky to implement because type-info can be normalized, but `sql-jdbc.sync/database-type->base-type` expects a non-normalized database-type.

File: modules/drivers/vertica/src/metabase/driver/vertica.clj
Comment (calherries) [inline modules/drivers/vertica/src/metabase/driver/vertica.clj:158]: I know this should have comments explaining this complex expression but that's not the focus of this PR.

**Comment by camsaul** [comment]
@calherries I opened https://github.com/metabase/metabase/issues/27250 to track disallowing this in Kondo altogether. It looks like adding these rules to the Kondo config is super simple. Do you want to take a crack at adding the rules and fixing other usages as part of this PR? Anyone using Metabase in Turkish would forever be in your debt

File: src/metabase/util/honeysql_extensions.clj
Comment (camsaul) [reply src/metabase/util/honeysql_extensions.clj:221]: 👍 I don't know why we didn't already have this function to be honest

File: src/metabase/driver/sql/query_processor.clj
Comment (camsaul) [inline src/metabase/driver/sql/query_processor.clj:158]:
```suggestion
  {:arglists '([driver unit field-or-value field-or-value]), :added "0.46.0"}
```

Since third-party driver authors are implementing Metabase drivers it's nice to add a little bit of extra metadata like this so driver authors can know when something was added just in case they want to support multiple versions in one driver

Comment (camsaul) [reply src/metabase/driver/sql/query_processor.clj:158]: You should also add a little blurb about this new method and the purpose of it in 

https://github.com/metabase/metabase/blob/master/docs/developers-guide/driver-changelog.md

File: src/metabase/driver/sql/query_processor.clj
Comment (camsaul) [inline src/metabase/driver/sql/query_processor.clj:157]: It's also nice to mention what the method is _for_ -- this method is used by the default implementation of `sql.qp/->honeysql` for `:datetime-diff`, or whatever -- and when or why people would want to implement it (implement it if your driver supports `:datetime-diff` and you want to use the default implementation of `->honeysql` for `:datetime-diff` clauses, which is recommended because ______ (built-in error-checking, etc.))

Comment (camsaul) [reply src/metabase/driver/sql/query_processor.clj:157]: I'm really not trying to be incredibly fussy here, but we have third-party driver authors and we should really error on the side of documenting our driver interface too much

File: modules/drivers/vertica/src/metabase/driver/vertica.clj
Comment (camsaul) [reply modules/drivers/vertica/src/metabase/driver/vertica.clj:158]: I would have been ok with those in this PR. Either way tho

**Review by camsaul — APPROVED** [review]
Cool. This is 1000 times easier to read I think. Left a few notes RE documenting the new `datetime-diff` method but you can go ahead and do that and then merge this. You don't need to wait for another review

File: src/metabase/driver/sql/query_processor.clj
Comment (calherries) [reply src/metabase/driver/sql/query_processor.clj]: Cool, updated the changelog https://github.com/metabase/metabase/commit/2e5f76df104cddddc8297b9121df4675f646100b

File: src/metabase/driver/sql/query_processor.clj
Comment (calherries) [reply src/metabase/driver/sql/query_processor.clj]: Updated https://github.com/metabase/metabase/pull/27237/commits/06eb4b6d880e0a310820f9faa6fdd272d37ea560

**Pull request merged**
